### PR TITLE
417 Websockets

### PR DESCRIPTION
**Pull request by DutchBen**

### CODE REVIEW

File: packages/orchestrator-ui-components/src/components/WfoBadges/WfoHeaderBadge/WfoHeaderBadge.tsx [hunk 1]
```diff
@@ -18,7 +18,6 @@ export const WfoHeaderBadge: FC<HeaderBadgeProps> = ({
         ? {
               height: 24,
               display: 'flex',
-              paddingLeft: 0,
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WfoBadges/WfoHeaderBadge/WfoHeaderBadge.tsx:21]: This padding needs to stay here, it makes the buttons look different
image

image

File: packages/orchestrator-ui-components/src/components/WfoPageTemplate/WfoPageHeader/WfoPageHeader.tsx [hunk 2]
```diff
@@ -27,6 +28,8 @@ export interface WfoPageHeaderProps {
     handleLogoutClick: () => void;
 }
 
+const WEBSOCKET_FEATURE_TOGGLE = true;
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WfoPageTemplate/WfoPageHeader/WfoPageHeader.tsx:31]: Think we want this stuck at false since it does not work currently, or how about to only show the badge when an actual env var is set?

File: packages/orchestrator-ui-components/src/messages/nl-NL.json [hunk 3]
```diff
@@ -12,7 +12,9 @@
         "tasks": "Taken",
         "title": "Workflow Orchestrator",
         "welcome": "Welkom",
-        "workflows": "Workflows"
+        "workflows": "Workflows",
+        "websocketConnected": "De websocket is aktief",
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/messages/nl-NL.json:16]: *actief

File: packages/orchestrator-ui-components/src/rtk/api.ts [hunk 4]
```diff
@@ -13,6 +13,11 @@ export enum BaseQueryTypes {
     custom = 'custom',
 }
 
+export enum CacheTags {
+    engineStatus = 'engineStatus',
+    testStatus = 'testStatus', // This is a placeholder for now. Having only one messes up the type of the tag in the onCacheEntryAdded function.
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/rtk/api.ts:18]: I think our linting rules do not apply to comments. Suggesting to place comments above what needs to be commented. Can can either be above `testStatus` or the entire enum I would say

File: packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+import { addToastMessage } from '@/rtk/slices/toastMessages';
+import type { RootState } from '@/rtk/store';
+import { ToastTypes } from '@/types';
+import { getToastMessage } from '@/utils/getToastMessage';
+
+import { CacheTags, orchestratorApi } from '../api';
+
+const getWebSocket = (url: string) => {
+    // TODO: Implement authentication taking this into account: https://stackoverflow.com/a/77060459o
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts:11]: Page not found, seems incorrect url

File: packages/orchestrator-ui-components/src/utils/getToastMessage.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import { ToastTypes } from '../types';
+import type { Toast } from '../types';
+
+export const getToastMessage = (
+    type: ToastTypes,
+    text: string, // We use string here instead of Toast['text'] because we want to prevent passing in react component because they trigger an "unsynchronizable values in payload detected" error',
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/utils/getToastMessage.ts:6]: Is that because the redux store cannot store the toast otherwise?

File: packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    return new WebSocket(url);
+};
+
+const PING_INTERVAL = 5000;
+const DEBOUNCE_CLOSE_INTERVAL = 10000;
+
+// From https://redux-toolkit.js.org/rtk-query/usage/customizing-queries#streaming-data-with-no-initial-request
+const streamMessagesApi = orchestratorApi.injectEndpoints({
+    endpoints: (build) => ({
+        streamMessages: build.query<boolean, void>({
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts:21]: This is quite some block of code, that is also hard to follow to be honest. I would like to suggest to see if there are ways to abstract away some things.

File: packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts (hunk 5 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts:11]: Fixed

File: packages/orchestrator-ui-components/src/components/WfoBadges/WfoWebsocketStatusBadge/WfoWebsocketStatusBadge.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+import { EuiToolTip } from '@elastic/eui';
+
+import { useOrchestratorTheme } from '@/hooks/useOrchestratorTheme';
+import { WfoStatusDotIcon, WfoXCircleFill } from '@/icons';
+import { useStreamMessagesQuery } from '@/rtk/endpoints/streamMessages';
+
+import { WfoHeaderBadge } from '../WfoHeaderBadge';
+
+``;
```
[the remark this answers is not in the document]
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/components/WfoBadges/WfoWebsocketStatusBadge/WfoWebsocketStatusBadge.tsx:13]: Removed

File: packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts (hunk 7 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts:21]: I've added more comments with explanation. It's hard and doesn't make it a lot clearer to abstract code away I think.

File: packages/orchestrator-ui-components/src/utils/getToastMessage.ts (hunk 6 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/utils/getToastMessage.ts:6]: I've only moved this code around so this was already there. I don't exactly remember but I think it was that redux does store it but you loose the ability to a do history playback in the redux tools which is nice when debugging.

**Review by ricardovdheijden — APPROVED** [review]

File: packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import { getToastMessage } from '@/utils/getToastMessage';
+
+import { CacheTags, orchestratorApi } from '../api';
+
+const getWebSocket = (url: string) => {
+    // TODO: Implement authentication taking this into account: https://stackoverflow.com/questions/4361173/http-headers-in-websockets-client-api/77060459#77060459
+    return new WebSocket(url);
+};
+
+const PING_INTERVAL = 5000;
```
Comment (tjeerddie) [inline packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts:15]: probably good to add the format to the naming `PING_INTERVAL_MS`

File: packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+                // Send a ping message every to the websocket server to keep the connection alive
+                const pingInterval = setInterval(() => {
+                    webSocket.send('ping');
+                }, PING_INTERVAL);
+
+                const debounceCloseWebSocket = debounce(() => {
+                    webSocket.close();
+                }, DEBOUNCE_CLOSE_INTERVAL);
+                // Start the debounced function to close the websocket when no 'pong' message is received after DEBOUNCE_CLOSE_INTERVAL
+                debounceCloseWebSocket();
```
Comment (tjeerddie) [inline packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts:76]: I guess this is a way, would be nicer to start a timeout after ping is send and clear it after pong is received.

File: packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+                webSocket.addEventListener(
+                    'message',
+                    (message: MessageEvent<string>) => {
+                        const tagOrPong = message.data.trim();
+
+                        if (tagOrPong === 'pong') {
+                            // Reset the debounced every time a 'pong' message is received
+                            debounceCloseWebSocket();
+                        } else {
```
Comment (tjeerddie) [inline packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts:86]: add a return in the if and remove the else, I find it unnecessary nesting :3

File: packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts (hunk 10 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts:76]: I feel this is more stable since you only start the interfal once.

File: packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts (hunk 11 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/rtk/endpoints/streamMessages.ts:86]: done

**Review by tjeerddie — APPROVED** [review]

File: packages/orchestrator-ui-components/src/components/WfoPageTemplate/WfoPageHeader/WfoPageHeader.tsx [hunk 12]
```diff
@@ -34,6 +35,8 @@ export interface WfoPageHeaderProps {
 
 const ENABLE_THEME_SWITCH =
     process.env.NEXT_PUBLIC_USE_THEME_TOGGLE === 'true' || false;
+const ENABLE_WEBSOCKET =
```
Comment (tjeerddie) [inline packages/orchestrator-ui-components/src/components/WfoPageTemplate/WfoPageHeader/WfoPageHeader.tsx:38]: add `NEXT_PUBLIC_USE_WEBSOCKET` to .env.example and add usage in readme

File: README.md [hunk 13]
```diff
@@ -20,6 +20,10 @@ npm run dev
 
 This makes the orchestrator ui run on http://localhost:3000
 
+## Websocket
+
+The application tries to open a websocket to ORCHESTRATOR_WEBSOCKET_URL that defaults to ws://localhost:8080 to get updates as they happen. The messages received through this endpoint are used to invalidate the frontend cache that triggers a refetch of data were needed.
+
```
Comment (tjeerddie) [inline README.md:26]: add usage of `NEXT_PUBLIC_USE_WEBSOCKET`
